revocation/ocsp/ocsp.go

```diff
@@ -0,0 +1,180 @@
+// Package OCSP provides methods for checking the OCSP revocation status of a certificate chain
+// as well as errors related to these checks
+package ocsp
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/x509"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+
+	"golang.org/x/crypto/ocsp"
+)
+
+type Options struct {
+	CertChain       []*x509.Certificate
+	SigningTime     time.Time
+	HTTPClient      *http.Client
+	MergeErrorsFunc func(errors []error) error
+}
+
+const pkixNoCheckOID string = "1.3.6.1.5.5.7.48.1.5"
+const ocspMaxResponseSize int64 = 20480 //bytes
+
+func OCSPStatus(opts Options) error {
+	for i, cert := range opts.CertChain {
+		if i != (len(opts.CertChain) - 1) {
+			if err := validateIssuer(cert, opts.CertChain[i+1]); err != nil {
+				return CheckOCSPError{Err: errors.New("invalid chain: expected chain to be correct and complete with each cert issued by the next in the chain")}
+			}
+		} else {
+			if err := validateIssuer(cert, cert); err != nil {
+				return CheckOCSPError{Err: errors.New("invalid chain: expected chain to end with root cert")}
+			}
+		}
+	}
+
+	wg := sync.WaitGroup{}
+	certResults := make([]error, len(opts.CertChain))
+	for i, cert := range opts.CertChain {
+		if i != (len(opts.CertChain) - 1) {
+			wg.Add(1)
+			// Assume cert chain is accurate and next cert in chain is the issuer
+			go certOCSPStatus(cert, opts.CertChain[i+1], &certResults[i], &wg, opts)
+		} else {
+			// Last is root cert, which will never be revoked by OCSP
+			certResults[len(opts.CertChain)-1] = nil
+		}
+	}
+
+	wg.Wait()
+	return opts.MergeErrorsFunc(certResults)
+}
+
+func validateIssuer(subject *x509.Certificate, issuer *x509.Certificate) error {
+	if err := subject.CheckSignatureFrom(issuer); err != nil {
+		return err
+	}
+	if !bytes.Equal(issuer.RawSubject, subject.RawIssuer) {
+		return errors.New("parent's subject does not match issuer for a cert in the chain")
+	}
+	return nil
+}
+
+func certOCSPStatus(cert *x509.Certificate, issuer *x509.Certificate, result *error, wg *sync.WaitGroup, opts Options) {
+	defer wg.Done()
+	var err error
+
+	ocspURLs := cert.OCSPServer
```

Comment:
OCSP urls can be ldap or other string, how are we handling that?

Reply:
In the spec, it said we are only supporting HTTP. Thus, it attempts to make an HTTP request to the specified location. If it is unable to reach it, then it returns an error. My initial hesitation was if a cert had a server set to a LDAP or string location that would translate to a potentially malicious site, however that appears to not be an issue since LDAP/string shouldn't be in a format that would translate to a usable HTTP location (and this appears to be something that is the responsibility of the CA, since there is no other way to indicate the type of location).

If there is hesitation, I can fast-track this error by parsing the server string for starting with "http" and return an error if it doesn't, but I'm not sure if all OCSP servers accessed via HTTP are required to start with http://

Reply:
Doing further research, these cannot be any string. They are required to be a URI, meaning they would need to specify a scheme. The only allowed URL scheme is HTTP, so I will validate that the string starts with http:// to fast-track the error.

Reference: https://www.rfc-editor.org/rfc/rfc6960
> CAs that support an OCSP service, either hosted locally or provided
   by an Authorized Responder, MUST provide for the inclusion of a value
   for a Uniform Resource Identifier (URI) [[RFC3986](https://www.rfc-editor.org/rfc/rfc3986)] accessLocation and
   the OID value id-ad-ocsp for the accessMethod in the
   AccessDescription SEQUENCE.

Reply:
Thats correct but [RFC3986](https://www.rfc-editor.org/rfc/rfc3986)] allow vast set of URI but Notation only support http and https. Can we please add validation for this and return proper error message `not supported` to customer?

Reply:
I can change the error message to convey that the specific protocol is unsupported. However, do we want to support HTTPS? This is not a recommended implementation for requesters or responders per [RFC5280](https://www.rfc-editor.org/rfc/rfc5280#section-8):
>When certificates include a cRLDistributionPoints extension with an https URI or similar scheme, circular dependencies can be introduced. The relying party is forced to perform an additional path validation in order to obtain the CRL required to complete the initial path validation! Circular conditions can also be created with an https URI (or similar scheme) in the authorityInfoAccess or subjectInfoAccess extensions.  At worst, this situation can create unresolvable dependencies.
>
>CAs SHOULD NOT include URIs that specify https, ldaps, or similar schemes in extensions.  CAs that include an https URI in one of these extensions MUST ensure that the server's certificate can be validated without using the information that is pointed to by the URI. Relying parties that choose to validate the server's certificate when obtaining information pointed to by an https URI in the cRLDistributionPoints, authorityInfoAccess, or subjectInfoAccess extensions MUST be prepared for the possibility that this will result in unbounded recursion.

This explicitly mentions CRL, but the same concept applies to OCSP. We would need to check the status of the SSL certificate for the HTTPS request. There is potential for unbounded recursion if the SSL is issued by the same CA (or falls into some circular chain). Additionally, certificates are public, there is no need for encryption. I think adding HTTPS support introduces a potential availability risk without adding much benefit to users (since this is specified as a SHOULD NOT in the RFC)

Reply:
Consensus after discussion was not supporting HTTPS and calling this out in the spec. Resolving this thread